protocols/jabber/sasl.c

```diff
@@ -142,15 +147,33 @@ xt_status sasl_pkt_mechanisms(struct xt_node *node, gpointer data)
 		/* The rest will be done later, when we receive a <challenge/>. */
 	} else if (sup_plain) {
 		int len;
+		GString *gs;
+		char *username;
 
-		xt_add_attr(reply, "mechanism", "PLAIN");
+		if (!want_hipchat) {
+			xt_add_attr(reply, "mechanism", "PLAIN");
+			username = jd->username;
+		} else {
+			username = jd->me;
+		}
+
+		/* set an arbitrary initial size to avoid reallocations */
+		gs = g_string_sized_new(128);
 
 		/* With SASL PLAIN in XMPP, the text should be b64(\0user\0pass) */
-		len = strlen(jd->username) + strlen(ic->acc->pass) + 2;
-		s = g_malloc(len + 1);
-		s[0] = 0;
-		strcpy(s + 1, jd->username);
-		strcpy(s + 2 + strlen(jd->username), ic->acc->pass);
+		g_string_append_c(gs, '\0');
+		g_string_append(gs, username);
+		g_string_append_c(gs, '\0');
+		g_string_append(gs, ic->acc->pass);
```

Comment:
hmmm, drafting it here

```
g_string_append_printf(gs, "%c%s%c%s", 0, username, 0, ic->acc->pass);

if (want_hipchat) {
    g_string_append_printf(gs, "%c%s", 0, set_getstr(&ic->acc->set, "resource"))
}
```

I dunno.


How about

```
char *pass = ic->acc->pass;
char *resource = set_getstr(&ic->acc->set, "resource");
if (want_hipchat) {
    g_string_printf(gs, "%c%s%c%s%c%s", 0, username, 0, pass, 0, resource);
} else {
    g_string_printf(gs, "%c%s%c%s", 0, username, 0, pass);
}
```

Maybe better? (I don't know if it works, though.)